File: apps/library/src/google_books.rs
use serde::{Deserialize, Serialize};
use std::time::Duration;

#[derive(Debug, Clone, Serialize)]
pub struct SearchHit {
    pub google_id: String,
    pub title: String,
    pub authors: String,
    pub isbn: Option<String>,
    pub cover_url: Option<String>,
}

#[derive(Deserialize)]
struct VolumesResponse {
    #[serde(default)]
    items: Vec<Volume>,
}

#[derive(Deserialize)]
struct Volume {
    id: String,
    #[serde(rename = "volumeInfo")]
    volume_info: VolumeInfo,
}

#[derive(Deserialize)]
struct VolumeInfo {
    title: Option<String>,
    #[serde(default)]
    authors: Vec<String>,
    #[serde(default, rename = "industryIdentifiers")]
    identifiers: Vec<Identifier>,
    #[serde(rename = "imageLinks")]
    image_links: Option<ImageLinks>,
}

#[derive(Deserialize)]
struct Identifier {
    #[serde(rename = "type")]
    kind: String,
    identifier: String,
}

#[derive(Deserialize)]
struct ImageLinks {
    thumbnail: Option<String>,
    #[serde(rename = "smallThumbnail")]
    small_thumbnail: Option<String>,
}

pub async fn search(query: &str, api_key: Option<&str>) -> Result<Vec<SearchHit>, String> {
    let trimmed = query.trim();
    if trimmed.is_empty() {
        return Ok(Vec::new());
    }
    let normalized: String = trimmed
        .chars()
        .filter(|c| !c.is_whitespace() && *c != '-')
        .collect();
    let is_isbn = matches!(normalized.len(), 10 | 13)
        && normalized
            .chars()
            .all(|c| c.is_ascii_digit() || c == 'X' || c == 'x');
    let query_str = if is_isbn {
        format!("isbn:{}", normalized.to_uppercase())
    } else {
        trimmed.to_string()
    };
    let q = urlencoding::encode(&query_str);
    let mut url = format!(
        "https://www.googleapis.com/books/v1/volumes?q={q}&maxResults=10&printType=books"
    );
    if let Some(key) = api_key {
        url.push_str(&format!("&key={}", urlencoding::encode(key)));
    }

    let client = reqwest::Client::builder()
        .timeout(Duration::from_secs(8))
        .user_agent("andromeda-library/0.1")
        .build()
        .map_err(|e| format!("client build: {e}"))?;

    let resp = client
        .get(&url)
        .send()
        .await
        .map_err(|e| format!("request: {e}"))?;

    if !resp.status().is_success() {
        return Err(format!("google books status {}", resp.status()));
    }

    let data: VolumesResponse = resp
        .json()
        .await
        .map_err(|e| format!("parse: {e}"))?;

    Ok(data
        .items
        .into_iter()
        .map(|v| {
            let info = v.volume_info;
            let isbn = pick_isbn(&info.identifiers);
            let cover_url = info
                .image_links
                .as_ref()
                .and_then(|l| l.thumbnail.clone().or_else(|| l.small_thumbnail.clone()))
                .map(|u| u.replacen("http://", "https://", 1));
            SearchHit {
                google_id: v.id,
                title: info.title.unwrap_or_else(|| "Untitled".to_string()),
                authors: info.authors.join(", "),
                isbn,
                cover_url,
            }
        })
        .collect())
}

fn pick_isbn(ids: &[Identifier]) -> Option<String> {
    ids.iter()
        .find(|i| i.kind == "ISBN_13")
        .or_else(|| ids.iter().find(|i| i.kind == "ISBN_10"))
        .map(|i| i.identifier.clone())
}

File: apps/library-go/google_books.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"time"
)

type SearchHit struct {
	GoogleID string  `json:"google_id"`
	Title    string  `json:"title"`
	Authors  string  `json:"authors"`
	ISBN     *string `json:"isbn,omitempty"`
	CoverURL *string `json:"cover_url,omitempty"`
}

type volumesResponse struct {
	Items []volume `json:"items"`
}

type volume struct {
	ID         string     `json:"id"`
	VolumeInfo volumeInfo `json:"volumeInfo"`
}

type volumeInfo struct {
	Title       string       `json:"title"`
	Authors     []string     `json:"authors"`
	Identifiers []identifier `json:"industryIdentifiers"`
	ImageLinks  *imageLinks  `json:"imageLinks"`
}

type identifier struct {
	Kind       string `json:"type"`
	Identifier string `json:"identifier"`
}

type imageLinks struct {
	Thumbnail      string `json:"thumbnail"`
	SmallThumbnail string `json:"smallThumbnail"`
}

func googleBooksSearch(ctx context.Context, query, apiKey string) ([]SearchHit, error) {
	trimmed := strings.TrimSpace(query)
	if trimmed == "" {
		return nil, nil
	}
	normalized := strings.Map(func(r rune) rune {
		if r == ' ' || r == '\t' || r == '\n' || r == '-' {
			return -1
		}
		return r
	}, trimmed)
	isISBN := (len(normalized) == 10 || len(normalized) == 13) && isISBNChars(normalized)
	q := trimmed
	if isISBN {
		q = "isbn:" + strings.ToUpper(normalized)
	}
	u := "https://www.googleapis.com/books/v1/volumes?q=" + url.QueryEscape(q) + "&maxResults=10&printType=books"
	if apiKey != "" {
		u += "&key=" + url.QueryEscape(apiKey)
	}

	client := &http.Client{Timeout: 8 * time.Second}
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, u, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("User-Agent", "andromeda-library-go/0.1")
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("request: %w", err)
	}
	defer resp.Body.Close()
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, fmt.Errorf("google books status %s", resp.Status)
	}
	var data volumesResponse
	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
		return nil, fmt.Errorf("parse: %w", err)
	}
	hits := make([]SearchHit, 0, len(data.Items))
	for _, v := range data.Items {
		title := v.VolumeInfo.Title
		if title == "" {
			title = "Untitled"
		}
		hit := SearchHit{
			GoogleID: v.ID,
			Title:    title,
			Authors:  strings.Join(v.VolumeInfo.Authors, ", "),
		}
		if isbn := pickISBN(v.VolumeInfo.Identifiers); isbn != "" {
			hit.ISBN = &isbn
		}
		if v.VolumeInfo.ImageLinks != nil {
			cover := v.VolumeInfo.ImageLinks.Thumbnail
			if cover == "" {
				cover = v.VolumeInfo.ImageLinks.SmallThumbnail
			}
			if cover != "" {
				cover = strings.Replace(cover, "http://", "https://", 1)
				hit.CoverURL = &cover
			}
		}
		hits = append(hits, hit)
	}
	return hits, nil
}

func pickISBN(ids []identifier) string {
	for _, i := range ids {
		if i.Kind == "ISBN_13" {
			return i.Identifier
		}
	}
	for _, i := range ids {
		if i.Kind == "ISBN_10" {
			return i.Identifier
		}
	}
	return ""
}

func isISBNChars(s string) bool {
	for _, c := range s {
		if (c < '0' || c > '9') && c != 'X' && c != 'x' {
			return false
		}
	}
	return true
}
